kale/pipeline/multi_domain_adapter.py

```diff
@@ -0,0 +1,314 @@
+
+        Returns:
+            torch.Tensor: divergence
+        """
+
+        moment_loss = 0
+        for i in range(self.k_moment):
+            moment_loss += losses._moment_k(x, domain_labels, i + 1)
+
+        return moment_loss
+
+
+class _DINTrainer(BaseMultiSourceTrainer):
+    def __init__(
+        self,
+        dataset,
+        feature_extractor,
+        task_classifier,
+        n_classes: int,
+        target_domain: str,
+        kernel: str = "linear",
+        kernel_mul: float = 2.0,
+        kernel_num: int = 5,
+        **base_params,
+    ):
+        """Domain independent network.
+
+        """
+        super().__init__(dataset, feature_extractor, task_classifier, n_classes, target_domain, **base_params)
+        self.kernel = kernel
+        self.n_domains = len(self.domain_to_idx.values())
+        self.classifier = task_classifier(self.feature_dim, n_classes)
+        self._kernel_mul = kernel_mul
+        self._kernel_num = kernel_num
+
+    def compute_loss(self, batch, split_name="V"):
+        x, y, domain_labels = batch
+        phi_x = self.forward(x)
+        loss_dist = self._compute_domain_dist(phi_x, domain_labels)
+        src_idx = torch.where(domain_labels != self.target_label)
+        tgt_idx = torch.where(domain_labels == self.target_label)
+        cls_output = self.classifier(phi_x)
+        loss_cls, ok_src = losses.cross_entropy_logits(cls_output[src_idx], y[src_idx])
+        _, ok_tgt = losses.cross_entropy_logits(cls_output[tgt_idx], y[tgt_idx])
+
+        task_loss = loss_cls
+        log_metrics = {
+            f"{split_name}_source_acc": ok_src,
+            f"{split_name}_target_acc": ok_tgt,
+            f"{split_name}_domain_acc": loss_dist,
+        }
+
+        return task_loss, loss_dist, log_metrics
+
+    def _compute_domain_dist(self, x, domain_labels):
+        if self.kernel == "linear":
+            kx = torch.mm(x, x.T)
+        else:
+            raise ValueError("Other kernels have not been implemented yet!")
+        domain_label_mat = one_hot(domain_labels, num_classes=self.n_domains)
+        domain_label_mat = domain_label_mat.float()
+        ky = torch.mm(domain_label_mat, domain_label_mat.T)
+        return losses.hsic(kx, ky, device=self.device)
+
+
+class MFSANTrainer(BaseMultiSourceTrainer):
+    def __init__(
+        self,
+        dataset,
+        feature_extractor,
+        task_classifier,
+        n_classes: int,
+        target_domain: str,
+        domain_feat_dim: int = 100,
+        kernel_mul: float = 2.0,
+        kernel_num: int = 5,
+        **base_params,
+    ):
+        """
+        Reference: Zhu, Y., Zhuang, F. and Wang, D., 2019, July. Aligning domain-specific distribution and classifier
+            for cross-domain classification from multiple sources. In Proceedings of the AAAI Conference on Artificial
+            Intelligence (Vol. 33, No. 01, pp. 5989-5996).
```

Comment:
"Vol. 33, No. 01" not needed for conferences, vol and no for journals typically.